src/Compile.hs

```diff
@@ -19,339 +19,538 @@ import           System.FilePath                ( replaceExtension
-     where
-      compileField :: Field -> String
-      compileField field = case field of
-        Field (name, exp) ->
-          " " <> name <> ": " <> compile astPath outputPath exp
-        FieldSpread exp -> " ..." <> compile astPath outputPath exp
-
-    FieldAccess record field ->
-      compile astPath outputPath record <> compile astPath outputPath field
-
-    JSExp content -> content
-
-    ListConstructor elems ->
-      "([" <> intercalate ", " (compileListItem <$> elems) <> "])"
-     where
-      compileListItem :: ListItem -> String
-      compileListItem li = case li of
-        ListItem   exp -> compile astPath outputPath exp
-        ListSpread exp -> " ..." <> compile astPath outputPath exp
-
-    TupleConstructor elems ->
-      "([" <> intercalate ", " (compile astPath outputPath <$> elems) <> "])"
-
-    Where exp (first : cs) ->
-      "((__x__) => {\n  "
-        <> compileIs first
-        <> concat (("  else " ++) . compileIs <$> cs)
-        -- TODO: Add an else for undefined patterns error and make it throw.
-        <> "})("
-        <> compile astPath outputPath exp
-        <> ")"
-     where
-      compilePattern :: String -> Pattern -> String
-      compilePattern _     (PVar _)  = "true"
-      compilePattern _     PAny      = "true"
-      compilePattern scope (PNum  n) = scope <> " === " <> n
-      compilePattern scope (PStr  n) = scope <> " === \"" <> n <> "\""
-      compilePattern scope (PBool n) = scope <> " === " <> n
-      compilePattern scope (PCon n)
-        | n == "String"  = "typeof " <> scope <> " === \"string\""
-        | n == "Boolean" = "typeof " <> scope <> " === \"boolean\""
-        | n == "Number"  = "typeof " <> scope <> " === \"number\""
-        | otherwise      = ""
-      compilePattern scope (PCtor n []) =
-        scope <> ".__constructor === " <> "\"" <> removeNamespace n <> "\""
-      compilePattern scope (PCtor n ps) =
-        scope
-          <> ".__constructor === "
-          <> "\""
-          <> removeNamespace n
-          <> "\""
-          <> if not (null args) then " && " <> args else ""
-       where
-        args =
-          intercalate " && "
-            $   filter (not . null)
-            $   compileCtorArg scope n
-            <$> zip [0 ..] ps
-      compilePattern scope (PRecord m) =
-        intercalate " && " $ filter (not . null) $ M.elems $ M.mapWithKey
-          (compileRecord scope)
-          m
-
-      compilePattern scope (PSpread pat) = compilePattern scope pat
-      compilePattern scope (PList   [] ) = scope <> ".length === 0"
-      compilePattern scope (PList items) =
-        scope
-          <> ".length "
-          <> lengthComparator items
-          <> " "
-          <> show (length items)
-          <> " && "
-          <> intercalate
-               " && "
-               (   (\(item, i) ->
-                     compilePattern (scope <> "[" <> show i <> "]") item
+  compile config e@(Solved _ area@(Area (Loc _ l _) _) exp) =
```

Comment:
I find that parsing this file is very hard and the blocks are very large. Is there a way we could break this function into more composite parts?

Reply:
Yeah, it's mainly big blocks of smaller blocks full of pattern matching. Probably we should move Compile to a separate folder and split things slowly. We could probably at least make a separate file for patterns as I think that's a piece where a lot of complexity is hidden and that will definitely grow bigger. So that would be the whole `where` stuff mostly.